test/router-luatest/reload_test.lua

```diff
@@ -0,0 +1,271 @@
+                if service == internal.static_router[service_name] then
+                    error('Service have not been reloaded yet')
+                end
+            end)
+        end
+    end, {service_name})
+end
+
+local function test_basic_template(router)
+    router:exec(function()
+        -- Basic test of callrw and callro
+        local bucket_id = 1
+        ivshard.router.callrw(bucket_id, 'insert', {'test', {1, bucket_id}})
+        ivshard.router.callrw(bucket_id, 'vshard.storage.sync', {})
+        local res, err = ivshard.router.callro(bucket_id, 'get', {'test', {1}})
+        ilt.assert_equals(err, nil)
+        ilt.assert_equals(res, {1, bucket_id})
+    end)
+end
+
+g.test_basic = function(g)
+    -- Latest meaningful commit:
+    --     "router: fix reload problem with global function refs".
+    local hash = '139223269cddefe2ba4b8e9f6e44712f099f4b35'
+    local router = create_router_at(hash)
+    test_basic_template(router)
+    reload_router(router)
+    test_basic_template(router)
+    vtest.drop_instance(g, router)
+end
+
+local function test_discovery_template(g, router)
+    router:exec(function(uuid)
+        local router = ivshard.router.internal.static_router
+        local discovery = router.discovery_service
+
+        -- Everything is all right
+        ivtest.service_wait_for_ok(discovery,
+            {on_yield = ivshard.router.discovery_wakeup})
+        ilt.assert_equals(router.known_bucket_count, 100)
+
+        -- Break connection and the number of known buckets
+        local conn = router.replicasets[uuid].master.conn
+        local known = router.known_bucket_count
+        router.replicasets[uuid].master.conn = nil
+        router.known_bucket_count = known - 1
+
+        ivtest.service_wait_for_error(discovery, 'Error during discovery',
+            {on_yield = ivshard.router.discovery_wakeup})
+
+        -- Restore everything
+        router.replicasets[uuid].master.conn = conn
+        router.known_bucket_count = known
+
+        -- It's all right again
+        ivtest.service_wait_for_ok(discovery,
+            {on_yield = ivshard.router.discovery_wakeup})
+    end, {g.replica_2_a:replicaset_uuid()})
+end
+
+g.test_discovery = function(g)
+    -- Service_info was introduced in the following commit,
+    -- the test cannot be run without it:
+    --     "router: add saving of background service statuses".
+    local hash = 'f5f386a5a35e6e5efd8f4f2ed1b3d208fdae9095'
+    local router = create_router_at(hash)
+    test_discovery_template(g, router)
+    reload_router(router, 'discovery_service')
+    test_discovery_template(g, router)
+    vtest.drop_instance(g, router)
+end
+
+local function test_master_search_template(g, router, auto_master_cfg)
+    vtest.router_cfg(router, auto_master_cfg)
+
+    -- Working with first replicaset (2 instances)
+    local rs_uuid, master_uuid = g.replica_1_a:exec(function()
+        local info = box.info()
+        return info.cluster.uuid, info.uuid
```

Comment:
Now we have to use `util.replicaset_uuid()`. Lets make a rebase and test on the latest tarantool core.

Reply:
As rebase was already done, I only fixed getting names. I propose to use `server:replicaset_uuid()` and `server:instance_uuid()` rather than `utill` in favor of code simplicity and readability. I know, that `box.info` is called twice in such situation, but it doesn't significantly affects the time test runs.